src/allencell_ml_segmenter/thresholding/thresholding_service.py

```diff
@@ -0,0 +1,153 @@
+
+from allencell_ml_segmenter.core.event import Event
+from allencell_ml_segmenter.core.file_input_model import FileInputModel
+from allencell_ml_segmenter.core.subscriber import Subscriber
+from allencell_ml_segmenter.main.experiments_model import ExperimentsModel
+from allencell_ml_segmenter.main.main_model import MainModel
+from allencell_ml_segmenter.thresholding.thresholding_model import (
+    ThresholdingModel,
+)
+from allencell_ml_segmenter.core.task_executor import (
+    NapariThreadTaskExecutor,
+    ITaskExecutor,
+)
+from allencell_ml_segmenter.main.viewer import IViewer
+from cyto_dl.models.im2im.utils.postprocessing.auto_thresh import AutoThreshold  # type: ignore
+
+
+class ThresholdingService(Subscriber):
+    def __init__(
+        self,
+        thresholding_model: ThresholdingModel,
+        experiments_model: ExperimentsModel,
+        file_input_model: FileInputModel,
+        main_model: MainModel,
+        viewer: IViewer,
+        task_executor: ITaskExecutor = NapariThreadTaskExecutor.global_instance(),
+    ):
+        super().__init__()
+        # Models
+        self._thresholding_model: ThresholdingModel = thresholding_model
+        self._experiments_model: ExperimentsModel = experiments_model
+        self._file_input_model: FileInputModel = file_input_model
+        self._main_model: MainModel = main_model
+
+        # napari viewer
+        self._viewer: IViewer = viewer
+
+        # Task Executor
+        self._task_executor: ITaskExecutor = task_executor
+
+        self._thresholding_model.subscribe(
+            Event.ACTION_THRESHOLDING_VALUE_CHANGED,
+            self,
+            self._on_threshold_changed,
+        )
+
+        self._thresholding_model.subscribe(
+            Event.ACTION_THRESHOLDING_AUTOTHRESHOLDING_SELECTED,
+            self,
+            self._on_threshold_changed,
+        )
+
+        self._thresholding_model.subscribe(
+            Event.ACTION_SAVE_THRESHOLDING_IMAGES,
+            self,
+            self._save_thresholded_images,
+        )
+
+    def _handle_thresholding_error(self, error: Exception) -> None:
+        show_info("Thresholding failed: " + str(error))
+
+    def _on_threshold_changed(self, _: Event) -> None:
+        # if we havent thresholded yet, keep track of original layers.
+        # need to check this on first threshold change, since user can add images
+        # between finishing prediction and starting thresholding
+        # if they are using images from a directory.
+        original_layers: Optional[OrderedDict[str, np.ndarray]] = (
+            self._thresholding_model.get_original_layers()
+        )
+        if original_layers is None:
+            self._thresholding_model.set_original_layers(
+                self._viewer.get_layers()
+            )
+
+        # Get layers to threshold.
+        # if there are segmentations displayed in the viewer, only threshold those images.
+        layers_to_threshold: OrderedDict[str, np.ndarray] = (
```

Comment:
Doesnt `_thresholding_model` need to be updated to have any layers that might have just been added to the viewer (drag n drop) ?

Reply:
Is `self._thresholding_model.get_original_layers()` the layerName -> layer dict that we discussed?

I was imagining the `thresholding_model` encapsulating this functionality, and exposing something like `addLayer(layerName, layer)`.